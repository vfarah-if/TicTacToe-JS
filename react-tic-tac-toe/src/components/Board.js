import React, { Component } from 'react';
import Square from './Square';

export const TOP_LEFT = 0;
export const TOP_MIDDLE = 1;
export const TOP_RIGHT = 2;
export const MIDDLE_LEFT = 3;
export const MIDDLE = 4;
export const MIDDLE_RIGHT = 5;
export const BOTTOM_LEFT = 6;
export const BOTTOM_MIDDLE = 7;
export const BOTTOM_RIGHT = 8;

class Board extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isPlayerOneActive: true,
            playerOne: 'X',
            playerTwo: '0',
            squares: Array(9).fill(null),
            winner: null
        };
    }

    handleClick(i) {
        if (!this.state.winner) {
            const squares = this.state.squares.slice();
            if (this.isEmpty(squares[i])) {
                squares[i] = this.state.isPlayerOneActive ? this.state.playerOne : this.state.playerTwo;
                const winner = this.calculateWinner(squares);
                this.updateState(squares, !this.state.isPlayerOneActive, winner);
            }
        }
    }

    isEmpty(square) {
        return square === null;
    }

    updateState(squares, isPlayerOneActive, winner) {
        this.setState({
            squares: squares,
            isPlayerOneActive: isPlayerOneActive,
            winner: winner
        });
    }

    renderSquare(i) {
        return <Square
            value={this.state.squares[i]}
            onClick={() => this.handleClick(i)} />;
    }

    calculateWinner(squares) {
        const lines = [
            // Horizontal Lines
            [TOP_LEFT, TOP_MIDDLE, TOP_RIGHT],
            [MIDDLE_LEFT, MIDDLE, MIDDLE_RIGHT],
            [BOTTOM_LEFT, BOTTOM_MIDDLE, BOTTOM_RIGHT],
            // Vertical Lines
            [TOP_LEFT, MIDDLE_LEFT, BOTTOM_LEFT],
            [TOP_MIDDLE, MIDDLE, BOTTOM_MIDDLE],
            [BOTTOM_RIGHT, MIDDLE_RIGHT, TOP_RIGHT]
            // Diagonal Lines

        ];
        for (let index = 0; index < lines.length; index++) {
            const [a, b, c] = lines[index];
            if (this.lineIsEqual(squares, a, b, c)) {
                return squares[a];
            }
        }

        return null;
    }

    lineIsEqual(squares, a, b, c) {
        return squares[a] && squares[a] === squares[b] && squares[a] === squares[c];
    }

    render() {
        const { isPlayerOneActive, playerOne, playerTwo, winner } = this.state;
        let status = `Next player: ${isPlayerOneActive ? playerOne : playerTwo}`;
        if (winner) {
            status = `Winner: ${winner}`;
        }
        return (
            <div>
                <div className="status">{status}</div>
                <div className="board-row">
                    {this.renderSquare(TOP_LEFT)}
                    {this.renderSquare(TOP_MIDDLE)}
                    {this.renderSquare(TOP_RIGHT)}
                </div>
                <div className="board-row">
                    {this.renderSquare(MIDDLE_LEFT)}
                    {this.renderSquare(MIDDLE)}
                    {this.renderSquare(MIDDLE_RIGHT)}
                </div>
                <div className="board-row">
                    {this.renderSquare(BOTTOM_LEFT)}
                    {this.renderSquare(BOTTOM_MIDDLE)}
                    {this.renderSquare(BOTTOM_RIGHT)}
                </div>
            </div>
        );
    }
}

export default Board;