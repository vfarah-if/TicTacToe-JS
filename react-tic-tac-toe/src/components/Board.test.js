import React from 'react';
import { shallow, mount } from 'enzyme';

import Board, { 
    TOP_LEFT, TOP_MIDDLE, TOP_RIGHT, 
    MIDDLE_LEFT, MIDDLE, MIDDLE_RIGHT 
    } 
    from './Board';

function getSquares(wrapper) {
    return wrapper.find('Square');
}

function getStatus(wrapper) {
    return wrapper.find('div.status');
}

describe('Board', () => {
    it('should set X or player one as the first player', () => {
        const wrapper = shallow(<Board/>)        
        const status = getStatus(wrapper);
        expect(status).toBeTruthy();
        expect(status.text()).toBe('Next player: X')
    });

    it('should create a board with 9 squares', () => {
        const wrapper = shallow(<Board/>)
        const elements = getSquares(wrapper);
        expect(elements.length).toBe(9)
    });    

    it('should toggle active players', () => {
        const wrapper = shallow(<Board/>)
        const elements = getSquares(wrapper);
        let status = getStatus(wrapper);
        expect(status.text()).toBe('Next player: X')
        const firstSquare = elements.first();
        firstSquare.simulate('click');
        status = getStatus(wrapper);
        expect(status.text()).toBe('Next player: 0')
    });    

    it('should prevent the same position being edited more than once', () => {
        const wrapper = mount(<Board/>)
        const squares = getSquares(wrapper);
        const firstSquare = squares.first();
        expect(firstSquare.text()).toBe('');
        expect(getStatus(wrapper).text()).toBe('Next player: X')

        firstSquare.simulate('click');
        wrapper.update();

        expect(firstSquare.text()).toBe('X');
        expect(getStatus(wrapper).text()).toBe('Next player: 0')

        firstSquare.simulate('click');
        wrapper.update();

        expect(firstSquare.text()).toBe('X');
        expect(getStatus(wrapper).text()).toBe('Next player: 0')
    }); 
    
    it('should present a winner when the top horizontal row is uniform', () => {
        const wrapper = mount(<Board/>)
        // X X X
        // 0 0 -
        // - - -
        wrapper.instance().handleClick(TOP_LEFT);   // X
        wrapper.instance().handleClick(MIDDLE_LEFT);// 0
        wrapper.instance().handleClick(TOP_MIDDLE); // X
        wrapper.instance().handleClick(MIDDLE);     // 0
        wrapper.instance().handleClick(TOP_RIGHT);  // X

        expect(getStatus(wrapper).text()).toBe('Winner: X')
    });

    it('should present a winner when the middle horizontal row is uniform', () => {
        const wrapper = mount(<Board/>)
        // 0 0 -
        // X X X
        // - - -
        wrapper.instance().handleClick(MIDDLE_LEFT);
        wrapper.instance().handleClick(TOP_LEFT);   
        wrapper.instance().handleClick(MIDDLE);     
        wrapper.instance().handleClick(TOP_MIDDLE); 
        wrapper.instance().handleClick(MIDDLE_RIGHT);

        expect(getStatus(wrapper).text()).toBe('Winner: X')
    });
    
});