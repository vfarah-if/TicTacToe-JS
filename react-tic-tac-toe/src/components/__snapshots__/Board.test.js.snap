// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should create a board with 9 configured squares using shallow render 1`] = `
Array [
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>,
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>,
  <div
    className="board-row"
  >
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
    <Square
      onClick={[Function]}
      value={null}
    />
  </div>,
]
`;
